Install without compiling gRPC module

* Operating System version: Amazon Linux 4.4.4
  * Firebase SDK version: 5.12.1
  * Library version: ?
  * Firebase Product: Messaging

#### Steps to reproduce:

`yarn add firebase-admin` or `npm install firebase-admin` will start to compile some gRPC native module with node-gyp since there aren't any pre-built binaries for this platform.

This requires a lot of memory, making it so we can't run our server on a t2.micro (512MB), and instead need to use t2.medium, just because it's trying to compile this native module.

Is there any way to use firebase messaging without needing to compile native module on install..?

_I worked around this by compiling it inside docker before deploying. It would be nice if these packages was split out as separate things so we didn't need to install everything including this native dependency when I just need the messaging parts_

I have a similar issue

npm install -g firebase-admin    

> grpc@1.13.1 install /usr/local/lib/node_modules/firebase-admin/node_modules/grpc
> node-pre-gyp install --fallback-to-build --library=static_library

node-pre-gyp WARN Using request for node-pre-gyp https download 
node-pre-gyp WARN Pre-built binaries not installable for grpc@1.13.1 and node@8.11.3 (node-v57 ABI, glibc) (falling back to source compile with node-gyp) 
node-pre-gyp WARN Hit error EACCES: permission denied, mkdir '/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/src/node' 
gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/8.11.3"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/.node-gyp"
gyp WARN install got an error, rolling back install
gyp WARN install got an error, rolling back install
gyp ERR! configure error 
gyp ERR! stack Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/.node-gyp'
gyp ERR! System Linux 4.15.0-1014-gcp
gyp ERR! command "/usr/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--library=static_library" "--module=/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/src/node/extension_binary/node-v57-linux-x64-glibc/grpc_node.node" "--module_name=grpc_node" "--module_path=/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/src/node/extension_binary/node-v57-linux-x64-glibc" "--napi_version=3" "--node_abi_napi=napi"
gyp ERR! cwd /usr/local/lib/node_modules/firebase-admin/node_modules/grpc
gyp ERR! node -v v8.11.3
gyp ERR! node-gyp -v v3.7.0
gyp ERR! not ok 
node-pre-gyp ERR! build error 
node-pre-gyp ERR! stack Error: Failed to execute '/usr/bin/node /usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --library=static_library --module=/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/src/node/extension_binary/node-v57-linux-x64-glibc/grpc_node.node --module_name=grpc_node --module_path=/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/src/node/extension_binary/node-v57-linux-x64-glibc --napi_version=3 --node_abi_napi=napi' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
node-pre-gyp ERR! stack     at emitTwo (events.js:126:13)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:925:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)
node-pre-gyp ERR! System Linux 4.15.0-1014-gcp
node-pre-gyp ERR! command "/usr/bin/node" "/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build" "--library=static_library"
node-pre-gyp ERR! cwd /usr/local/lib/node_modules/firebase-admin/node_modules/grpc
node-pre-gyp ERR! node -v v8.11.3
node-pre-gyp ERR! node-pre-gyp -v v0.10.3
node-pre-gyp ERR! not ok 
Failed to execute '/usr/bin/node /usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --library=static_library --module=/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/src/node/extension_binary/node-v57-linux-x64-glibc/grpc_node.node --module_name=grpc_node --module_path=/usr/local/lib/node_modules/firebase-admin/node_modules/grpc/src/node/extension_binary/node-v57-linux-x64-glibc --napi_version=3 --node_abi_napi=napi' (1)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! grpc@1.13.1 install: `node-pre-gyp install --fallback-to-build --library=static_library`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the grpc@1.13.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2018-08-06T09_18_08_653Z-debug.log

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,8 @@
 # Unreleased
 
 - [changed] Upgraded Cloud Firestore client to v0.16.0.
+- [changed] Firestore and Storage client libraries are now defined as optional
+  dependencies.
 
 # v5.13.1
 

--- a/package.json
+++ b/package.json
@@ -54,12 +54,14 @@
   "dependencies": {
     "@firebase/app": "^0.3.1",
     "@firebase/database": "^0.3.1",
-    "@google-cloud/firestore": "^0.16.0",
-    "@google-cloud/storage": "^1.6.0",
-    "@types/google-cloud__storage": "^1.1.7",
     "@types/node": "^8.0.53",
     "jsonwebtoken": "8.1.0",
     "node-forge": "0.7.4"
+  },
+  "optionalDependencies": {
+    "@google-cloud/firestore": "^0.16.0",
+    "@google-cloud/storage": "^1.6.0",
+    "@types/google-cloud__storage": "^1.1.7"
   },
   "devDependencies": {
     "@types/chai": "^3.4.34",

--- a/src/firestore/firestore.ts
+++ b/src/firestore/firestore.ts
@@ -118,7 +118,8 @@ function initFirestore(app: FirebaseApp): Firestore {
     throw new FirebaseFirestoreError({
       code: 'missing-dependencies',
       message: 'Failed to import the Cloud Firestore client library for Node.js. '
-          + 'Make sure to install the "@google-cloud/firestore" npm package.',
+          + 'Make sure to install the "@google-cloud/firestore" npm package. '
+          + `Original error: ${err}`,
     });
   }
   return new firestoreDatabase(options);

--- a/src/firestore/firestore.ts
+++ b/src/firestore/firestore.ts
@@ -110,7 +110,16 @@ export function getFirestoreOptions(app: FirebaseApp): any {
 
 function initFirestore(app: FirebaseApp): Firestore {
   const options = getFirestoreOptions(app);
-  // Lazy-load the Firestore implementation here, which in turns loads gRPC.
-  const firestoreDatabase: typeof Firestore = require('@google-cloud/firestore');
+  let firestoreDatabase: typeof Firestore;
+  try {
+    // Lazy-load the Firestore implementation here, which in turns loads gRPC.
+    firestoreDatabase = require('@google-cloud/firestore');
+  } catch (err) {
+    throw new FirebaseFirestoreError({
+      code: 'missing-dependencies',
+      message: 'Failed to import the Cloud Firestore client library for Node.js. '
+          + 'Make sure to install the "@google-cloud/firestore" npm package.',
+    });
+  }
   return new firestoreDatabase(options);
 }

--- a/src/storage/storage.ts
+++ b/src/storage/storage.ts
@@ -65,7 +65,7 @@ export class Storage implements FirebaseServiceInterface {
       throw new FirebaseError({
         code: 'storage/missing-dependencies',
         message: 'Failed to import the Cloud Storage client library for Node.js. '
-          + 'Make sure to install the "@google-cloud/storage" npm package.'
+          + 'Make sure to install the "@google-cloud/storage" npm package. '
           + `Original error: ${err}`,
       });
     }

--- a/src/storage/storage.ts
+++ b/src/storage/storage.ts
@@ -60,13 +60,13 @@ export class Storage implements FirebaseServiceInterface {
 
     let storage: typeof gcs;
     try {
-      /* tslint:disable-next-line:no-var-requires */
       storage = require('@google-cloud/storage');
-    } catch (e) {
+    } catch (err) {
       throw new FirebaseError({
         code: 'storage/missing-dependencies',
         message: 'Failed to import the Cloud Storage client library for Node.js. '
-          + 'Make sure to install the "@google-cloud/storage" npm package.',
+          + 'Make sure to install the "@google-cloud/storage" npm package.'
+          + `Original error: ${err}`,
       });
     }
 

--- a/tslint-test.json
+++ b/tslint-test.json
@@ -8,7 +8,7 @@
       "variables-before-functions"
     ],
     "prefer-object-spread": false,
-    "no-implicit-dependencies": [true, "dev"],
+    "no-implicit-dependencies": [true, "dev", "optional"],
     "max-classes-per-file": false,
     "max-line-length": [true, 120],
     "no-consecutive-blank-lines": false,

--- a/tslint.json
+++ b/tslint.json
@@ -8,7 +8,7 @@
       "variables-before-functions"
     ],
     "prefer-object-spread": false,
-    "no-implicit-dependencies": [true, "dev"],
+    "no-implicit-dependencies": [true, "dev", "optional"],
     "max-classes-per-file": false,
     "max-line-length": [true, 120],
     "no-consecutive-blank-lines": false,

